internal/scheduler/legacyscheduler.go

```diff
@@ -318,151 +400,180 @@ func uuidFromUlidString(ulid string) (uuid.UUID, error) {
+		if limit.Cmp(q) == 1 {
 			return false, fmt.Sprintf(
-				"job requests 0 %s, but the minimum is %s",
-				resourceType, limit.String(),
+				"job requests %s %s, but the minimum is %s",
+				q.String(), resourceType, limit.String(),
 			)
 		}
 	}
 	return true, ""
 }
 
-func (sched *LegacyScheduler) selectNodeForPod(ctx context.Context, jobId uuid.UUID, job *api.Job, bind bool) (*PodSchedulingReport, error) {
-	podSpec := util.PodSpecFromJob(job)
-	if podSpec == nil {
-		return nil, errors.New("failed to get pod spec")
+type CandidateGangIterator struct {
+	ctx context.Context
+	CandidateJobsIterator
+	// Jobs are grouped into gangs by this annotation.
+	gangIdAnnotation string
+	// Jobs in a gang specify the number of jobs in the gang via this annotation.
+	gangCardinalityAnnotation string
+	// Map from queue to the gangs for that queue.
+	// Each gang is made up of a slice of jobs with equal value for gangIdAnnotation.
+	gangsByQueue map[string]map[string][]*JobSchedulingReport
+}
+
+func NewQueueCandidateGangIterator(ctx context.Context, it *CandidateJobsIterator, gangIdAnnotation, gangCardinalityAnnotation string) *CandidateGangIterator {
+	return &CandidateGangIterator{
+		ctx:                       ctx,
+		CandidateJobsIterator:     *it,
+		gangIdAnnotation:          gangIdAnnotation,
+		gangCardinalityAnnotation: gangCardinalityAnnotation,
+		gangsByQueue:              make(map[string]map[string][]*JobSchedulingReport),
 	}
+}
 
-	// Try to find a node for this pod.
-	// Store the report returned by the NodeDb.
-	req := schedulerobjects.PodRequirementsFromPodSpec(podSpec, sched.SchedulingConfig.Preemption.PriorityClasses)
-	var report *PodSchedulingReport
-	var err error
-	if bind {
-		report, err = sched.NodeDb.SelectAndBindNodeToPod(jobId, req)
-	} else {
-		report, err = sched.NodeDb.SelectNodeForPod(jobId, req)
+func (it *CandidateGangIterator) Next() ([]*JobSchedulingReport, error) {
+	// Once this function has returned error,
+	// it will return this error on every invocation.
+	if it.err != nil {
+		return nil, it.err
 	}
-	if err != nil {
-		return nil, err
+
+	// Get one job at a time from the underlying iterator until we either
+	// 1. get a job that isn't part of a gang, in which case we yield it immediately, or
+	// 2. get the final job in a gang, in which case we yield the entire gang.
+	for {
+		report, err := it.CandidateJobsIterator.Next()
+		if err != nil {
+			return nil, err
+		}
+		if report == nil {
+			return nil, nil
+		}
+
+		gangId, gangCardinality, isGangJob, err := GangIdAndCardinalityFromJob(report.Job, it.gangIdAnnotation, it.gangCardinalityAnnotation)
+		if err != nil {
+			log := ctxlogrus.Extract(it.ctx)
+			logging.WithStacktrace(log, err).Errorf("failed to get gang cardinality for job %s", report.Job.Id)
+			gangCardinality = 1 // Schedule jobs with invalid gang cardinality one by one.
+		}
+		if isGangJob {
+			jobsByGangId, ok := it.gangsByQueue[report.Job.Queue]
```

Comment:
I wonder if it's possible to write a default multimap now we have generics, in which case all this would reduce to a single line.  Isuspectwe probably can't due to maps (and slices) being special cased int  he compiler.

Reply:
Doing so isn't necessary anymore, since we collect gangs on a per-queue basis now.